service/raft/raft_address_map.hh

```diff
@@ -164,13 +173,90 @@ class raft_address_map {
         }
     }
 
+    template <std::invocable<raft_address_map&> F>
+    void replicate(F f, std::experimental::source_location l = std::experimental::source_location::current()) {
+        if (_gate.is_closed()) {
+            return;
+        }
+
+        _replication_fiber = _replication_fiber.then([this, f = std::move(f), l] () -> future<> {
+            if (_gate.is_closed()) {
+                co_return;
+            }
+
+            auto holder = _gate.hold();
+            try {
+                co_await this->container().invoke_on_others([f] (raft_address_map& self) {
+                    if (self._gate.is_closed()) {
+                        return;
+                    }
+
+                    f(self);
+                });
+            } catch (...) {
+                rslog.error("raft_address_map::replicate (called from {}) failed: {}",
+                            l.function_name(), std::current_exception());
+            }
+        });
+    }
+
+    void replicate_set_nonexpiring_entry(const raft::server_id& id, const gms::inet_address& addr) {
+        replicate([id, addr] (raft_address_map& self) {
+            self.handle_set_nonexpiring_entry(id, addr);
+        });
+    }
+
+    void replicate_set_expiring_flag(const raft::server_id& id) {
+        replicate([id] (raft_address_map& self) {
+            self.handle_set_expiring_flag(id);
+        });
+    }
+
+    void handle_set_nonexpiring_entry(const raft::server_id& id, const gms::inet_address& addr) {
+        auto [it, _] = _map.try_emplace(id, addr);
+        auto& entry = it->second;
+
+        // Don't allow to remap to a different address
+        if (entry._addr != addr) {
+            on_internal_error(rslog, format(
+                "raft_address_map::handle_set_nonexpiring_entry: expected to get inet_address {} for raft server id {} (got {})",
+                entry._addr, id, addr));
+        }
+
+        if (entry.expiring()) {
+            entry._lru_entry = nullptr;
+        }
+    }
+
+    void handle_set_expiring_flag(const raft::server_id& id) {
```

Comment:
Perhaps we should use a single function with 3 params - ip, addr, expiring - that would handle all cases? It would coerce the entry on the destination shard into the parameter values passed to it.

Reply:
I prefer them separate for now.